### DISCUSSION TITLE
Design/general question on dragging features

### DISCUSSION THREAD

**Discussion by jameslavery** [Mapsui/Mapsui#2230] [Q&A] [answered]
I'm going to want to display a set of navigation waypoints on the map, and allow the user to drag them around individually.

I assume they should all be in one "Waypoints" layer. For efficiency though, if I want to allow dragging of a single waypoint, I assume I'll have to refresh the entire layer as it moves, which sounds inefficient.

Or am I misunderstanding the architecture?

Is there documentation on how dragging works? I assume EditManager is involved.

I see #2147 adds a way to drag an entire feature, which is exactly what I want, I think. I just need to work out how it works!

**Comment by pauldendulk** [comment, marked as answer]
Yes, the entire canvas needs to be redrawn. We could think of more efficient ways but for the most important scenarios, like dragging the map and mouse/pinch zooming, we need to redraw the entire map anyway, because every pixel changes. It might only help in a static situation, where you are looking at a vehicle moving on the map. So, this is not on the roadmap and I think it should not be giving all other things we want to work on.

Then again there is caching of the 'drawable'. When rendering we first turn a feature into a SkiaSharp object which is drawn onto the canvas in a later stage. This caching is a problem for features that change, like something which is dragged over the map. To prevent caching your layer need to be derived from IModifyFeatureLayer, like WritableLayer. 

btw, there is a PR underway in which we can avoid caching on the level of the individual feature. https://github.com/Mapsui/Mapsui/pull/2138, but I think this will go in v5.

On how dragging is supposed to work. Don't think of it of something fancy. You can do it any way you like. The EditManager was initially written as a sample and was later moved to Mapsui.NTS but it was not designed to be a core component. Perhaps it is easier if you copy it to your own project, or just look at the implementation and write your own logic based on it.

Earlier you asked for more guidance and documentation. A general principle I like to follow is that it should be clear how the basics work and with that understanding developers should be able to build their own things on top of that. So, what I think we should *not* strive for is to have a well designed, well documented EditManager which solves all your problems. Instead developers should understand how a feature is drawn to the map and should understand events in which they can change a feature. In this respect the documentation (and the code) should be improved.

**Reply by jameslavery** [reply]
Thanks, that's really useful.

I'm going to be rendering my own Features with SkiaSharp - I'll check out IModifyFeatuireLayer to see if I should be using that anyway.

Regarding dragging. Thanks for the pointer on the EditManager - sounds like I can/should roll my own solution.

The main thing I'm trying to work out is how to process the drag event if I'm dragging on a Feature. I've already got 'selection' of a Feature working by detecting tap on it. So once it's selected, presumably I need to process the mouse movement to drag rather than pan the map. Would the new Feature Drag support added by @cecildelakers be helpful here?

BTW the flip side of all this is that if we're successul, my colleague and I should be able to contribute support for Transverse Mercator projections. 😀

**Reply by pauldendulk** [reply]
About Transverse Mercator, Note, that we have support for many coordinate system in Mapsui.Extensions through DotSpatial.Projections.

**Reply by jameslavery** [reply]
Thanks - are you saying that Transverse Mercator is already supported, or that this is how we can add support for it? I'm assuming the latter.

**Reply by pauldendulk** [reply]
We support the projections that are supported by DotSpatial.Projections. I do not know exactly which projections it supports but I guess Transvere Mercator is supported.

**Reply by pauldendulk** [reply]
Then again, I guess it does make sense to add Transverse Mercator as a native Mapsui projection, since it is such an important one. I am a bit hesitant because this could evolve into the development of an entire projection library which we should perhaps not take on, but weighing pros and cons I guess it would be good to add it.

**Reply by jameslavery** [reply]
We're just checking whether DotSpatial.Projections supports Transverse Mercator - not clear at the moment.

We've got handling of Transverse Mercator working and running in a finished product in a SwiftUI project, and have ported it over to C# - and all its standalone tests work.

So first off we (my team of me and my colleague Bill) need to work out how to make it work with Mapsui.Then we could formalise it and bring it into Mapsui as a 'native' projection if we all agree.

**Reply by pauldendulk** [reply]
Ok, for the 'native' projection look at Projection.cs. BTW, somebody also contributed a Mercator.cs. This could probably also be integrated into Projection.cs.

**Reply by jameslavery** [reply]
Thanks - will do. Will let you know how we get on.
